src/urllib3/contrib/emscripten/fetch.py

```diff
@@ -288,6 +300,104 @@ def send(self, request: EmscriptenRequest) -> EmscriptenResponse:
             )
 
 
+class _JSPIReadStream(io.RawIOBase):
+    """A read stream that uses pyodide.ffi.run_sync to read from a Javascript fetch
+    response. This requires support for WebAssembly Javascript Promise Integration
+    in the containing browser, and for pyodide to be launched via runPythonAsync.
+    """
+
+    def __init__(
+        self,
+        js_read_stream: Any,
+        timeout: float,
+        request: EmscriptenRequest,
+        response: EmscriptenResponse,
+        js_abort_controller: Any,  # javascript AbortController for timeouts
+    ):
+        """Stream to read data from a Javascript fetch response
+
+        Args:
+            js_read_stream (Any): The Javascript stream reader
+            timeout (float): Timeout in seconds
+            request (EmscriptenRequest): The request we're handling
+            response (EmscriptenResponse): The response this stream is in
+            js_abort_controller (Any): A javascript AbortController object
+        """
+        self.js_read_stream = js_read_stream
+        self.timeout = timeout
+        self._is_closed = False
+        self._is_done = False
+        self.request: EmscriptenRequest | None = request
+        self.response: EmscriptenResponse | None = response
+        self.current_buffer = None
+        self.current_buffer_pos = 0
+        self.js_abort_controller = js_abort_controller
+
+    def __del__(self) -> None:
+        self.close()
+
+    # this is compatible with _base_connection
+    def is_closed(self) -> bool:
+        return self._is_closed
+
+    # for compatibility with RawIOBase
+    @property
+    def closed(self) -> bool:
+        return self.is_closed()
+
+    def close(self) -> None:
+        if not self.is_closed():
+            self.read_len = 0
+            self.read_pos = 0
+            self.js_read_stream = None
+            self._is_closed = True
+            self._is_done = True
+            self.request = None
+            self.response = None
+            super().close()
```

Comment:
We should probably call `self.js_read_stream.cancel()` when we're done with it. Also a nit: it reads a bit nicer with an early return.
```suggestion
    def close(self) -> None:
        if self.is_closed():
            return
        self._is_closed = True
        self._is_done = True
        self.read_len = 0
        self.read_pos = 0
        self.js_read_stream.cancel()
        self.js_read_stream = None
        self.request = None
        self.response = None
        super().close()
```